presto-native-execution/presto_cpp/main/operators/PartitionAndSerialize.cpp

```diff
@@ -68,19 +73,135 @@ class PartitionAndSerializeOperator : public Operator {
+      FlatVector<int32_t> partitionVector) {
+    auto numInput = input_->size();
+    auto insertInto = numInput;
+
+    for (auto forPartition = 0; forPartition < numPartitions_; ++forPartition) {
+      for (auto copyFromIdx = 0; copyFromIdx < numInput; ++copyFromIdx) {
+        // Copy only if copyFromIdx has null partition key and it is not already
+        // assigned to forPartition.
+        // If Partition key is null, copy the serialised data into target index
+        if (nullRows_.isValid(copyFromIdx) &&
+            partitions_[copyFromIdx] != forPartition) {
+          dataVector.setNoCopy(insertInto, dataVector.valueAt(copyFromIdx));
+          partitionVector.set(insertInto, forPartition);
+          insertInto++;
+        }
+      }
+    }
+    return insertInto;
+  }
+
+  // Identify the first non-empty element and replicate on all empty partitions
+  vector_size_t replicateAnyInEmptyPartitions(
+      FlatVector<int32_t>& partitionsVector,
+      FlatVector<StringView>& dataVector,
+      const vector_size_t start) {
+    auto rawPartitions = partitionsVector.mutableRawValues();
+    const auto numInput = input_->size();
+    std::set<int32_t> nonEmptyPartitions(
+        rawPartitions, rawPartitions + numInput);
+    StringView firstNonEmpty;
+    for (int idx = 0; idx < numInput; ++idx) {
+      if (!nullRows_.isValid(idx)) {
+        firstNonEmpty = dataVector.valueAt(idx);
+        break;
+      }
+    }
+
+    int32_t finalOutputSize = start;
+    for (auto idx = 0; idx < numPartitions_; ++idx) {
+      if (nonEmptyPartitions.find(idx) == nonEmptyPartitions.end()) {
+        rawPartitions[finalOutputSize] = idx;
+        dataVector.setNoCopy(finalOutputSize, firstNonEmpty);
+        finalOutputSize++;
+      }
+    }
+    return finalOutputSize;
+  }
+
+  // A partition key is considered null only if its recursive children all are
+  // nulls. collectNullRows will flatten the key columns and store the null bits
+  // in a SelectivityVector nullRows_
+  void collectNullRows() {
+    auto size = input_->size();
+    // if replicateNullsAndAny_ is false, we don't need to calcualte nullRows_
+    if (!replicateNullsAndAny_) {
+      nullRows_.resize(size);
+      nullRows_.clearAll();
+      return;
+    }
+    rows_.resize(size);
+    rows_.setAll();
+
+    nullRows_.resize(size);
+    nullRows_.clearAll();
+
+    decodedVectors_.resize(keyChannels_.size());
+
+    for (auto partitionKey : keyChannels_) {
+      auto& keyVector = input_->childAt(partitionKey);
+      if (keyVector->mayHaveNulls()) {
+        decodedVectors_[partitionKey].decode(*keyVector, rows_);
+        if (auto* rawNulls = decodedVectors_[partitionKey].nulls()) {
+          bits::orWithNegatedBits(
+              nullRows_.asMutableRange().bits(), rawNulls, 0, size);
+        }
+      }
+    }
+    nullRows_.updateBounds();
+  }
+
   RowVectorPtr getOutput() override {
     if (!input_) {
       return nullptr;
     }
+    collectNullRows();
```

Comment:
```
if (replicateNullsAndAny_ && numPartitions_ > 1) {
  collectNullRows();
}
```

Reply:
I did it this way to avoid branching in `getOutput` function, and the function `collectNullRows` already handles this condition along with initialising the null selectivity vectors appropriately.